src/main/java/com/google/devtools/build/lib/remote/merkletree/DirectoryTreeBuilder.java

```diff
@@ -140,6 +142,18 @@ private static int buildFromActionInputs(
                   metadataProvider.getMetadata(input),
                   "missing metadata for '%s'",
                   input.getExecPathString());
+          // A TreeArtifact's metadata has type REGULAR_FILE, not DIRECTORY, and thus requires
```

Comment:
Consider moving this to `build()` where we create (parent) directories. So this function can keep the precondition that all inputs are not tree artifact nor middleman.